examples_tests/01.HelloWorld/main.cpp

```diff
@@ -21,39 +28,282 @@ inline void debugCallback(nbl::video::E_DEBUG_MESSAGE_SEVERITY severity, nbl::vi
+	core::smart_refctd_ptr<video::CVulkanPhysicalDevice> gpu = nullptr;
+
+	uint32_t graphicsFamilyIndex(~0u);
+	uint32_t presentFamilyIndex(~0u);
+
+	// Todo(achal): Probably want to put these into some struct
+	nbl::video::ISurface::SFormat surfaceFormat;
+	nbl::video::ISurface::E_PRESENT_MODE presentMode;
+	nbl::video::ISurface::E_SURFACE_TRANSFORM_FLAGS preTransform;
+	nbl::asset::E_SHARING_MODE imageSharingMode;
+	using QueueFamilyIndicesArrayType = core::smart_refctd_dynamic_array<uint32_t>;
+	QueueFamilyIndicesArrayType queueFamilyIndices;
+
+	for (size_t i = 0ull; i < gpus.size(); ++i)
+	{
+		// Todo(achal): Hacks, get rid
+		gpu = core::smart_refctd_ptr_static_cast<video::CVulkanPhysicalDevice>(*(gpus.begin() + i));
+		auto vk_surface = core::smart_refctd_ptr_static_cast<video::ISurfaceVK>(surface);
+
+		bool isGPUSuitable = false;
+
+		// Check if the physical device has queue families which support both graphics and presenting, not necessarily overlapping
+		{
+			const auto& queueFamilyProperties = gpu->getQueueFamilyProperties();
+
+			for (uint32_t familyIndex = 0u; familyIndex < queueFamilyProperties.size(); ++familyIndex)
+			{
+				const auto& familyProperty = queueFamilyProperties.begin() + familyIndex;
+				if (familyProperty->queueFlags & video::IPhysicalDevice::E_QUEUE_FLAGS::EQF_GRAPHICS_BIT)
+					graphicsFamilyIndex = familyIndex;
+
+				if (surface->isSupported(gpu.get(), familyIndex))
+					presentFamilyIndex = familyIndex;
+
+				if ((graphicsFamilyIndex != ~0u) && (presentFamilyIndex != ~0u))
+				{
+					isGPUSuitable = true;
+					break;
+				}
+			}
+		}
+
+		// Check if this physical device supports the swapchain extension
+		// Todo(achal): Eventually move this to somewhere inside the engine
+		{
+			// Todo(achal): Get this from the user
+			const uint32_t requiredDeviceExtensionCount = 1u;
+			const char* requiredDeviceExtensionNames[] = { VK_KHR_SWAPCHAIN_EXTENSION_NAME };
 
+			uint32_t availableExtensionCount;
+			vkEnumerateDeviceExtensionProperties(gpu->getInternalObject(), NULL, &availableExtensionCount, NULL);
+			std::vector<VkExtensionProperties> availableExtensions(availableExtensionCount);
+			vkEnumerateDeviceExtensionProperties(gpu->getInternalObject(), NULL, &availableExtensionCount, availableExtensions.data());
+
+			bool requiredDeviceExtensionsAvailable = false;
+			for (uint32_t i = 0u; i < availableExtensionCount; ++i)
+			{
+				if (strcmp(availableExtensions[i].extensionName, requiredDeviceExtensionNames[0]) == 0)
+				{
+					requiredDeviceExtensionsAvailable = true;
+					break;
+				}
+			}
+
+			if (!requiredDeviceExtensionsAvailable)
+				isGPUSuitable = false;
+		}
+
+		// Check if the surface is adequate
+		{
+			std::vector<video::ISurface::SFormat> surfaceFormats =
+				gpu->getAvailableFormatsForSurface(surface.get());
+
+			std::vector<video::ISurface::E_PRESENT_MODE> presentModes =
+				gpu->getAvailablePresentModesForSurface(surface.get());

```

Comment:
could I get a bitfield instead of an `std::vector` (since there are only like 5 present modes)?

Also can you solve #145 while you're at it?

Reply:
I believe, by this, you want me to make changes here 
https://github.com/Devsh-Graphics-Programming/Nabla/blob/744bcea6f2630ab2a760d406dd62192148d0136f/examples_tests/01.HelloWorld/main.cpp#L239

i.e. based on an input value in the range `{-1, 0, 1}` choose a present mode 

and while I'm at it I think it would be wise to incorporate this as a factor in finding a suitable physical device, in other words, if we don't find a physical device with specified VSync then we didn't find a physical device at all

Reply:
I want you to return a `uint16_t` typed enum ( and change the enum for presentation modes to be ascending bitmasks), which is a bitwise OR of the supported presentation modes corresponding enums

w.r.t. device selection (which orthogonal to my comment here), prefer the device which has the same bit present as your desired mode, otherwise just pick whatever.

#145 is just about having OpenGL spoof Vulkan, so report it supports presentation modes (FIFO,FIFO_RELAXED and that other one which means no VSync) outlined in #145 and just set the correct glx/wgl swapInterval to a value of {-1,0,1} depending on the presentation mode chosen.

Reply:
> I want you to return a uint16_t typed enum ( and change the enum for presentation modes to be ascending bitmasks), which is a bitwise OR of the supported presentation modes corresponding enums

yes ofcourse

> #145 is just about having OpenGL spoof Vulkan, so report it supports presentation modes (FIFO,FIFO_RELAXED and that other one which means no VSync) outlined in #145 and just set the correct glx/wgl swapInterval to a value of {-1,0,1} depending on the presentation mode chosen.

so it is not meant for me?

Reply:
I think I'll assign this to you

Reply:
resolved